compiler/rustc_infer/src/traits/mod.rs

```diff
@@ -105,6 +109,65 @@ pub enum FulfillmentErrorCode<'tcx> {
     CodeAmbiguity,
 }
 
+pub struct ObligationsDedup<'a, 'tcx, T> {
+    obligations: &'a mut Vec<Obligation<'tcx, T>>,
+}
+
+impl<'a, 'tcx, T: 'tcx> ObligationsDedup<'a, 'tcx, T>
+where
+    T: Clone + Hash + Eq,
+{
+    pub fn from_vec(vec: &'a mut Vec<Obligation<'tcx, T>>) -> Self {
+        ObligationsDedup { obligations: vec }
+    }
+
+    pub fn extend<'b>(&mut self, iter: impl ExactSizeIterator<Item = Cow<'b, Obligation<'tcx, T>>>)
+    where
+        'tcx: 'b,
+    {
+        // obligation tracing has shown that initial batches added to an empty vec do not
+        // contain any duplicates, so there's no need to attempt deduplication
+        if self.obligations.is_empty() {
+            *self.obligations = iter.into_iter().map(Cow::into_owned).collect();
+            return;
+        }
+
+        let initial_size = self.obligations.len();
+        let current_capacity = self.obligations.capacity();
+        let iter = iter.into_iter();
+        let expected_new = iter.len();
```

Comment:
Just confused as to why `iter.into_iter()` is necessary here?

Reply:
Just a leftover, will remove.